### ISSUE TITLE
[Issue] Unclear how to set AZURE_COSMOS_CONNECTION_STRING for local sample running

### ISSUE THREAD

**Issue by pamelafox** [Azure/azure-dev#1127]
This is about this sample:
https://github.com/Azure-Samples/todo-python-mongo-swa-func

I am attempting to run the Azure function in a Github codespace, _before_ deploying it. 

**Describe the bug**

This isn't a bug, but a commentary on what could be made clearer.

The readme says:
"Before running, set the `AZURE_COSMOS_CONNECTION_STRING` environment variable to the connection-string for mongo/cosmos."

However, it doesn't give any tips for setting that string. Should I:
1) Deploy it to Azure and use the actual connection string?
2) Set up a Cosmos emulator in the Codespace somehow?
3) Set up a Mongo instance in the Codespace? (Maybe how I set up PostGres in my samples?)

**To Reproduce**

If you don't set the string, then you'll get an error when trying to run the function.

To see that error,

 * Open in Codespace
 * Run Task: Start Function.

"Exception: ConfigurationError: Empty host (or extra comma in host list)."

**Expected behavior**

I was hoping for a hint about where/how to get that string. I haven't used Mongo or Cosmos much so it's not immediately clear to me.

**Environment**
Information on your environment: Edge, Github Codespaces

**Additional context**

I can't run this sample locally due to both lack of M1 support for azure-functions-worker and lack of M1 support for devcontainers, so any guidance would need to work in Codespaces.

**Comment by jongio** [comment]
Could you do a PR to all templates with that text?  I'm thinking adding it to the readme.  Thanks!

**Comment by pamelafox** [comment]
I don't actually know the answer to how we should set the string, do you have a recommendation of the 3 options I listed?

1. Deploy it to Azure and use the actual connection string?
2. Set up a Cosmos emulator in the Codespace somehow?
3. Set up a Mongo instance in the Codespace? (Maybe how I set up PostGres in my samples?)

Happy to update READMEs once I'm confident in the best approach.

**Comment by jongio** [comment]
Assigning to myself to have a look.

**Comment by pamelafox** [comment]
Aaron recommended this blog post if the goal is to set up a local CosmosDB in devcontainer:
https://www.aaron-powell.com/posts/2022-08-24-improved-local-dev-with-cosmosdb-and-devcontainers/
Specific example at https://github.com/aaronpowell/end-to-end-graphql-workshop

**Comment by weikanglim** [comment]
@pamelafox The ToDo templates currently require an Azure connection, so currently you would want to:
1. Run `azd provision`.
2. After provision is successful, F5 / Start Function experience will now work. The app uses the deployed Cosmos DB.

Under the covers, `azd` will set the variable `AZURE_KEY_VAULT_ENDPOINT` in the environment file from the bicep output file. This is used by the application to download secrets from the KeyVault and initialize `AZURE_COSMOS_CONNECTION_STRING`. 

Definitely agree that the docs are not great in conveying this, and we would need to update this. The `readme.md` in `src/api` is written without `azd` being in the picture.

**Comment by pamelafox** [comment]
Okay, so that would mean you would need to first do a provision before being able to run locally? When I develop an app, I like to be able to run fully locally first, which is why I thought a local CosmosDB would be the approach, but apparently some of my DevRel colleague do use the cloud DB locally as you suggest. That scares me, personally, as I'd rarely use a prod DB locally. Perhaps the README could provide instructions for both approaches.

**Comment by jongio** [comment]
Yes, provision is required to run it locally because of the KV connection.  We should fix that though.  Can you file a new issue? Keep this one open for this specific issue,

New one is something like: I should be able to run the template locally w/o a provision.

**Comment by pamelafox** [comment]
Sure, https://github.com/Azure/azure-dev/issues/1183

**Comment by vhvb1989** [comment]
Here's one suggestion to avoid using the PROD resources from Azure during testing/debug/dev

```bash
# First, create a dev env. Choose a name like `my-dev` when azd ask for env name
azd up
# At this point, you can consider the app running as your staging/dev
# Do as many changes/test/debug to it. After changes call 
azd provision # for changes to infra
azd deploy # for changes to the apps only
# Once you get the initial template in a way that is ready to be a PROD app:
azd env new my-prod
# This would create a new resource group to hold the PROD-ready app.
# you'll get the links for your customers.
azd up
# switch back to your dev env to continue working on new features.
azd env use my-dev
# If no changes to the infrastructure, call only deploy to push new changes for your app
azd deploy
# Finally, switch back to my-prod whenever you want to publish to prod.
```

Note:  
This is just one strategy.  And it could be improved a little bit by calling `azd pipeline config` on each azd environment.
This would delegate the `provision+deploy` actions to a pipeline in GitHub or Azdo for every git commit that is pushed.
The only important detail for this improvement is to create once source repo for dev and another for prod.
The above code would change to:

```bash
# First, create a dev env. Choose a name like `my-dev` when azd ask for env name
azd up
# At this point, you can consider the app running as your staging/dev
# Do as many changes/test/debug to it. 
# Delegate updates to a pipeline. Create a new GitHub/Azdo repo
# by default, if there is not a git remote already, we would create the `origin` for the dev-env
azd pipeline config
# after any change, just commit and push your changes to `origin`
# Once you get the initial template in a way that is ready to be a PROD app:
azd env new my-prod
# Now, instead of running azd up for PROD, let's create a new git-remote for this PROD env
# note how we set the --remote-name to ask azd to add and set up a new remote
azd pipeline config --remote-name prod
# that's it. once azd pipeline finishes, select yes to pushing changes to the repo to start the pipeline
# The pipeline will create the PROD app
# switch back to your dev env to continue working on new features.
azd env use my-dev
# Now, after every change, just push changes to `origin` to update DEV
git push origin
# And to publish to PROD, push to `prod` remote
git push prod
```

**Comment by savannahostrowski** [comment]
Moving to Gallium

**Comment by seesharprun** [comment]
In the official Azure Cosmos DB quickstart AZD templates (for Microsoft Learn), we use post-provision hooks to "write back" the endpoint to an .env file or some other local mechanism.

See https://github.com/azure-samples/cosmos-db-nosql-nodejs-quickstart or https://github.com/azure-samples/cosmos-db-nosql-dotnet-quickstart for examples.

We also take your deploying account's principal ID and grant RBAC data-plane access to read/write data.

That's one way to solve this. Unfortunately, I haven't found a clean way to handle this in Python like I would with .NET (User Secrets) or Node.js (.env).
